rust/wal3/src/gc.rs

```diff
@@ -0,0 +1,313 @@
+            children.push(Self::drop_fragment(frag, &mut Setsum::default())?);
+        }
+        for snap in snapshot.snapshots.iter() {
+            children.push(
+                Box::pin(Self::drop_snapshot(
+                    storage,
+                    prefix,
+                    snap,
+                    throttle,
+                    snapshots,
+                    &mut Setsum::default(),
+                ))
+                .await?,
+            );
+        }
+        Ok(GarbageAction::DropSnapshot {
+            path_to_snapshot: snapshot.path.clone(),
+            snapshot_setsum: snapshot.setsum,
+            children,
+        })
+    }
+
+    #[allow(clippy::result_large_err)]
+    async fn replace_snapshot<C: SnapshotCache>(
+        storage: &Storage,
+        prefix: &str,
+        ptr: &SnapshotPointer,
+        throttle: &ThrottleOptions,
+        snapshots: &C,
+        first_to_keep: LogPosition,
+        drop_acc: &mut Setsum,
+    ) -> Result<GarbageAction, Error> {
+        let snapshot = match snapshots.get(ptr).await? {
+            Some(snapshot) => snapshot,
+            None => match Snapshot::load(throttle, storage, prefix, ptr).await? {
+                Some(snapshot) => snapshot,
+                None => {
+                    return Err(Box::new(ScrubError::MissingSnapshot {
+                        reference: ptr.clone(),
+                    })
+                    .into())
+                }
+            },
+        };
+        let mut ret_snapshot = snapshot.clone();
+        let mut ret_children = vec![];
+        for frag in std::mem::take(&mut ret_snapshot.fragments).into_iter() {
+            if frag.limit <= first_to_keep {
+                ret_children.push(Self::drop_fragment(&frag, drop_acc)?);
+                ret_snapshot.setsum -= frag.setsum;
+            } else {
+                ret_snapshot.fragments.push(frag);
+            }
+        }
+        for snap in std::mem::take(&mut ret_snapshot.snapshots).into_iter() {
+            if snap.limit <= first_to_keep {
+                ret_children.push(
+                    Self::drop_snapshot(storage, prefix, &snap, throttle, snapshots, drop_acc)
+                        .await?,
+                );
+                ret_snapshot.setsum -= snap.setsum;
+            } else if (snap.start..snap.limit).contains(&first_to_keep) {
+                let drop_acc_preserved = *drop_acc;
+                let GarbageAction::ReplaceSnapshot {
+                    old_path_to_snapshot,
+                    old_snapshot_setsum,
+                    new_snapshot,
+                    children,
+                } = Box::pin(Self::replace_snapshot(
+                    storage,
+                    prefix,
+                    &snap,
+                    throttle,
+                    snapshots,
+                    first_to_keep,
+                    drop_acc,
+                ))
+                .await?
+                else {
```

Comment:
don't understand this. can't it happen that a snapshot within this is NOT a candidate for GC?